src/mscorlib/shared/System/Globalization/DateTimeFormat.cs

```diff
@@ -1060,101 +1113,204 @@ internal static StringBuilder FormatStringBuilder(DateTime dateTime, ReadOnlySpa
 
             if (format.Length == 1)
             {
-                switch (format[0])
+                format = ExpandPredefinedFormat(format, ref dateTime, ref dtfi, ref offset);
+            }
+
+            return FormatCustomized(dateTime, format, dtfi, offset, result: null);
+        }
+
+        // Roundtrippable format. One of
+        //   012345678901234567890123456789012
+        //   ---------------------------------
+        //   2017-06-12T05:30:45.7680000-07:00
+        //   2017-06-12T05:30:45.7680000Z           (Z is short for "+00:00" but also distinguishes DateTimeKind.Utc from DateTimeKind.Local)
+        //   2017-06-12T05:30:45.7680000            (interpreted as local time wrt to current time zone)
+        private static bool TryFormatO(DateTime dateTime, TimeSpan offset, Span<char> destination, out int charsWritten)
+        {
+            const int MinimumBytesNeeded = 27;
+
+            int charsRequired = MinimumBytesNeeded;
+            DateTimeKind kind = DateTimeKind.Local;
+
+            if (offset == NullOffset)
+            {
+                kind = dateTime.Kind;
+                if (kind == DateTimeKind.Local)
                 {
-                    case 'O':
-                    case 'o':
-                        return FastFormatRoundtrip(dateTime, offset);
-                    case 'R':
-                    case 'r':
-                        return FastFormatRfc1123(dateTime, offset, dtfi);
+                    offset = TimeZoneInfo.Local.GetUtcOffset(dateTime);
+                    charsRequired += 6;
+                }
+                else if (kind == DateTimeKind.Utc)
+                {
+                    charsRequired += 1;
                 }
+            }
+            else
+            {
+                charsRequired += 6;
+            }
 
-                format = ExpandPredefinedFormat(format, ref dateTime, ref dtfi, ref offset);
+            if (destination.Length < charsRequired)
+            {
+                charsWritten = 0;
+                return false;
             }
+            charsWritten = charsRequired;
+
+            // Hoist most of the bounds checks on destination.
+            { var unused = destination[MinimumBytesNeeded - 1]; }
```

Comment:
Interesting trick. @AndyAyersMS, the JIT can exclude bounds checks within loops. Any way for it to optimize out the bounds checks for multiple uses outside of a loop, such as this?

Reply:
(This is just ported from the Utf8Formatter implementation.)

Reply:
Sure, if the array length is not known, and the code is writing at fixed indices, write or access the largest first, and the jit should know the subsequent smaller indices are in bounds. More or less what is done here.

If indices contain unknown values the jit can sometimes reason about similar patterns, but not always.